text/0000-pod.md

```diff
@@ -0,0 +1,195 @@
+remain.)
+
+Lastly, a plain old data type allows for the following simple implementation of
+a random value generator:
+
+```rust
+pub trait RandomValueGenerator {
+    fn generate_random_bytes(&mut self, bytes: &mut [u8]);
+
+    fn generate<T: Pod>(&mut self) -> T {
+        let mut t = mem::zeroed();
+        self.create_random_bytes(mem::as_mut_bytes(&mut t));
+        t
+    }
+}
+```
+
+## Various convenience functions using plain old data
+
+This section contains a list of useful functions that can be written with a
+plain old data trait.
+
+```rust
+/// Creates an object that has all bytes set to zero.
+fn zeroed<T: Pod>() -> T;
+```
+
+```rust
+/// Returns the mutable in-memory representation of an object.
+fn as_mut_bytes<T: Pod>(val: &mut T) -> &mut [u8];
+```
+
+```rust
+/// Turns a slice into a reference to a Pod type if it's suitable.
+///
+/// = Remarks
+///
+/// The buffer is suitable if it is large enough to hold the type and properly
+/// aligned.
+pub fn from_bytes<T: Pod>(buf: &[u8]) -> Option<&T>;
+
+pub fn from_mut_bytes<T: Pod>(buf: &mut [u8]) -> Option<&mut T>;
+```
+
+```rust
+impl<T> [T] {
+    /// Returns a mutable byte slice covering the same range as the slice.
+    pub fn as_mut_bytes(&mut self) -> &mut [u8] where T: Pod;
+}
+```
+
+# Detailed design
+[design]: #detailed-design
+
+Add a `pod` lang item and safe `Pod` trait to mark types which are valid when
+they contain arbitrary bit patterns.
+
+The following types are **Pod candidates**:
+
+* `u8`, `u16`, `u32`, `u64`, `usize`,
+* `i8`, `i16`, `i32`, `i64`, `isize`,
+* `f32`, `f64`,
+* raw pointers,
+* arrays of `Pod` types,
+* tuples of `Pod` types, and
+* structs where all fields are public and `Pod`.
+
+Only types that are Pod candidates can implement the `Pod` trait. Arrays
+of `Pod` types and tuples of `Pod` types automatically implement the `Pod`
+trait.
+
+Structs with `Drop` implementations are not Pod candidates since the drop flag
+is considered a private field.
+
+## Justification of the set of Pod candidates
+
+It is easy to see that the list above contains only plain old data types by
+applying the following criterion recursively:
+
+>If an object containing an arbitrary bit pattern can be constructed in safe
+>code today, then the type of the object is a plain old data type.†
+
+(† This excludes padding between struct fields.)
+
+It is also clear that references (including slices) and enums are not plain old
+data types. The case of structs with `Drop` implementations or hidden fields is
+somewhat harder. First of all, consider the following type:
+
+```rust
+struct Slice<T> {
+    ptr: *const T,
+    len: usize,
+}
+```
+
+The inherent methods of this type will likely make use of the fact that the user
+cannot modify the `len` field in order to guarantee safety. Since the compiler
+cannot know if the private fields must satisfy certain invariants, an
+implementation of the safe (!) `Pod` trait cannot be allowed for such types.
+
```

Comment:
I don't agree with the reasoning w r t allowing safe `Pod` for types with private internals. If a struct is declared `Pod`, then it's an error to assume that there are private invariants. (A struct might want to keep its internals private for other reasons, such as being able to rename the fields without breaking the API.)